tensorbay/cli/log.py

```diff
@@ -158,10 +166,232 @@ def _get_full_log(commit: Commit, branch_name: Optional[str]) -> str:
+        self.printing_pointer = self.new_print_pointer
+        self.sorted_leaves[self.printing_pointer] = node
+
+    def get_graph_info(self) -> Tuple[int, int, int, int]:
+        """Get the logging graph information.
+
+        Returns:
+            The logging graph information.
+
+        """
+        return (
+            self.printing_amount,
+            self.printing_pointer,
+            self.old_print_pointer,
+            self.new_print_pointer,
+        )
+
+
+def _get_graph_commits(
+    commit_id_to_branches: DefaultDict[str, List[str]],
+    sorted_leaves: List[_CommitNode],
+    printer: Callable[[Commit, Optional[str], _GraphStack], str],
+) -> Iterator[str]:
+    graph_stack = _GraphStack(sorted_leaves)
+    while True:
+        current_node = graph_stack.get_log_node()
+        parent = current_node.parent
+        if parent in graph_stack.sorted_leaves:
+            # Merge branch.
+            graph_stack.merge_branches(parent)
+
+        print_branch_name = ", ".join(commit_id_to_branches.get(current_node.commit.commit_id, ()))
+        yield printer(
+            current_node.commit,
+            print_branch_name,
+            graph_stack,
+        )
+        if not parent:
+            break
+        graph_stack.set_next_node(parent)
+
+
+def _add_graph(
+    func: Callable[[Commit, Optional[str]], str]
+) -> Callable[[Commit, Optional[str], _GraphStack], str]:
+    def wrapper(commit: Commit, branch_name: Optional[str], graph_stack: _GraphStack) -> str:
+        log = func(commit, branch_name)
+        (
+            printing_amount,
+            printing_pointer,
+            old_print_pointer,
+            new_print_pointer,
+        ) = graph_stack.get_graph_info()
+        split_lines = log.splitlines()
+        prefix = ["|"] * printing_amount
+        # Don't merge branches.
+        if old_print_pointer == -1:
+            details_prefix = " ".join(prefix)
+            details = "".join(
+                map(lambda item: f"{details_prefix} {item}\n", islice(split_lines, 1, None))
+            )
+            title_prefix = _get_title_prefix(printing_pointer, prefix)
+            return f"{title_prefix} {split_lines[0]}\n{details}"
+
+        # Merge branches.
+        prefix.append("|")
+        title_prefix = _get_title_prefix(printing_pointer, prefix)
+        merge_prefix = _get_merge_graph(printing_amount, new_print_pointer, old_print_pointer)
+        # Show one line commit.
+        if len(split_lines) == 1:
+            merge_prefix_graph = "\n".join(merge_prefix)
+            return f"{title_prefix} {split_lines[0]}\n{merge_prefix_graph}\n"
+
+        # Show complete commit.
+        details = "\n".join(_combine_details(merge_prefix, split_lines[1:]))
+        return f"{title_prefix} {split_lines[0]}\n{details}\n"
+
+    return wrapper
+
+
+def _get_merge_graph(
+    printing_amount: int, printing_pointer: int, old_print_pointer: int
+) -> List[str]:
+    graph_prefix = ["|", " "] * printing_amount
```

Comment:
will it be better to use `graph_prefix == ["| "] * printing_amount` ?

Reply:
`temp_prefix[index * 2 - 1] = "/"`, in this line, the `" "` after `"|"` will be replaced by `"/"`.

Reply:
can we use `"|/"` to replace?